### ISSUE TITLE
Sequence type (map/collect - dynamic DAG construction)

### ISSUE THREAD

**Issue by mgasner** [dagster-io/dagster#462]
When a ("mapper"?) solid has an output of Seq[T], with length N, the entire sub-DAG descending from that output, until a single terminal ("reducer"?) node with input(s) of types Seq[T], Seq[S], etc., is reached, should be executed N times, once for each member of the sequence. These mapped sub-DAGs should be nestable, provided the nested sub-DAG terminates before the outer sub-DAG. (I.e., parentheses must match.) I think that by requiring a single terminal node for the entire sub-DAG we don't lose expressiveness -- If there are any number of terminal nodes, nothing is lost by having them pass their outputs to a no-op reduction solid.

**Comment by mgasner** [comment]
I guess to really handle the streaming case, we really want to be able to yield the outputs in the sequence over time. You could think of a Kinesis listener, for instance, as a solid with no inputs and an output of type Seq[T], which waits for the stream and then yields repeatedly, each time triggering execution of a copy of the descendent sub-DAG. It's not clear to me how to conceive of the reducer here.

**Comment by mgasner** [comment]
#699

**Comment by mgasner** [comment]
[naiad_sosp2013.pdf](https://github.com/dagster-io/dagster/files/2797163/naiad_sosp2013.pdf)

**Comment by mgasner** [comment]
Reupping this per @beaucronin

**Comment by mgasner** [comment]
```Alexander Kiselyov 1:08 PM
Hi all! I'm currently studying Dagster and trying to use it for tidying up ML training. I got kind of stuck with the following problem. A complex solid ("train") in a pipeline needs to be run several times with somewhat different input, but the number of these runs is determined by an integer value in a global config (ML folds).  I see only three solutions:
• generate the pipeline on the fly based on an original config: more code = more bugs, difficult to use tools like Dagit, seems like an abuse of the framework;
• iterate over inputs in a single complex solid "train": looks good in Dagit, but limited parallelism;
• just generate the pipeline with a fixed number of parallel "train" solids (i.e. a fixed maximum number of runs), then use optional outputs: seems sort of ugly and with artificial constants.```

**Comment by newturok** [comment]
In Airflow there is also no clear way to do it: https://stackoverflow.com/questions/41517798/proper-way-to-create-dynamic-workflows-in-airflow

**Comment by mgasner** [comment]
it's not straightforward :)

**Comment by jamescasbon** [comment]
See also https://docs.prefect.io/core/concepts/mapping.html#prefect-approach

**Comment by bradleyfay** [comment]
Metaflow has a similar mapping concept: https://docs.metaflow.org/metaflow/basics#foreach

**Comment by schrockn** [comment]
@jamescasbon and @bradleyfay both of those examples in the documentation are generating a DAG from data that is known as graph construction time, rather than from data being produced at run time. If you want that capability that actually relatively straightforward to do by programmatically generating pipelines and composites

**Comment by TMiguelT** [comment]
Metaflow's foreach actually doesn't require you to have the data at construction time:

> Alternatively, you may want to branch based on data dynamically. This is the use case for a foreach branch.

Something like that would be ideal in dagster.

**Comment by JohnMav** [comment]
Shared this with @mgasner on slack: My team had a use case for this where we are retrieving a set of urls of unknown size and for each url we wanted to retrieve, parse, and upload the data. While we could do this all in one step we were hoping to be able isolate errors and prevent the whole process from being blocked if one url resulted in malformed data. e.g.

```python
@solid
def get_urls(_):
	url_list = pd.read_csv('urls_to_update.csv')
    # This can be a list of size 0-100
	for url in url_list:
		yield Output(url, "url")

@solid 
def parse_data(_, url):
	res = requests.get(url).json()
	structured_data = dict(
		url=url,
		col_a=res["col_a"],
		col_b=res["col_b"]	
	)
	yield Output(structured_data, 'result')

@solid
def upsert_data(_, data):
	# Some logic to load the data into warehouse 

@pipeline
def my_dynamic_pipeline():
	upsert_data(
		data=parse_data(
			url=get_urls()
		)
	)
```

**Comment by dbendelman** [comment]
This is pretty important for us as well. Currently we're forced to wrap the solids that need to run for each output as an independent pipeline external to the current one, which gets launched from inside the mapper solid for each output item. Also required us to build up custom machinery for launching pipelines over graphql because of #2674. Somewhat painful workaround and our dagit's Runs page is filled with lots of unnecessary pipelines. Would be great if there was first class support for this.

**Comment by w-bonelli** [comment]
To chime in, this would also be super useful for us (in particular the ability to map solids over some number of inputs not known until runtime). We're using the Dask executor and there's currently no easy way to make dynamically demonstrated pipelines `reconstructable`, since pipeline ctor functions need to have no parameters and live at module scope.

To get this working with the current version of Dagster, @mgasner suggested either:

- writing DAG shape out to a file and reading it in the `reconstructable` function, or
- custom code to apportion underlying resources to "waiting" pipelines

For now we are doing this directly in Dask, but would be awesome if Dagster supported this :)

**Comment by hmakait** [comment]
I'm currently investigating dagster vs. prefect and good support for this is pretty much a hard requirement for our use case.

I'd love to pick the implementation up myself, provided somebody volunteers to provide guidance since this would be my first contribution to dagster.

**Comment by mgasner** [comment]
@w-bonelli just wanted to make sure you've seen the new `build_reconstructable_pipeline` API

**Comment by w-bonelli** [comment]
@mgasner awesome! I had not, thanks for the heads up

**Comment by catherinewu** [comment]
For those who want to follow along / give feedback / share use cases, @alangenfeld will be posting updates to the implementation here: https://github.com/dagster-io/dagster/discussions/3295

**Comment by alangenfeld** [comment]
Support for a "map" step via `DynamicOutput` was released in an experimental state in `0.10.0`, with front-end support in `dagit` coming soon in `0.10.3`. Please send any feedback on this early version here or over in our slack.

A "collect" step involves some more invasive changes so is still incoming. Go ahead and 👍 this comment to voice your desire for "collect" support.

**Comment by alangenfeld** [comment]
`collect` support went out in `0.11.0`

Going to close this task out - feel free to file specific issues for feedback & bugs
